Fix MetadataValidator

bernt-matthias opened the pull request:
If no `check` or `skip` attribute is given, `self.check` and `self.skip` need to be initialized to `None`. Otherwise validation with `value.missing_meta(check=self.check, skip=self.skip)` is called with `value.missing_meta(check=[""], skip=[""])`.

This led to the stalling observed in https://github.com/galaxyproject/galaxy/issues/13136

Explanation: each data input parameter automatically gets a `MetadataValidator`. Since this validator was dysfunctional the job was scheduled anyway, but did not execute.

Appears that the validator might have been dysfunctional for a long time (since `check` and `skip` have been introduced here:  https://github.com/bernt-matthias/galaxy/commit/b9e242e4eb39f7d2d582008affc7ec694cbe170b)

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
fix MetadataValidator

If no check or skip is given, self.check and self.skip
need to be initialised to None. Otherwise validation with
`value.missing_meta(check=self.check, skip=self.skip)`
is called with `value.missing_meta(check=[""], skip=[""])`.

This led to the stalling observed in
https://github.com/galaxyproject/galaxy/issues/13136

Explanation: each data input parameter automatically gets
a MetadataValidator. Since this validator was dysfunctional
the job was scheduled anyway, but did not execute.

bernt-matthias committed:
fix MetadataValidator tests

disable the MetadataValidator that is added to data params
by default by adding `no_validation="true"` to the test parameters.
(see https://github.com/galaxyproject/galaxy/blob/bc939ca4e0f0404a20d2cd3fbcc922768c195c08/lib/galaxy/tools/parameters/basic.py#L1898)

Since the default is initialised with MetadataValidator() the message
was None. Now message is also initialised in __init__.

Analoguous change in DatasetOkValidator, UnspecifiedBuildValidator.

nsoranzo commented on lib/galaxy/tools/parameters/validation.py:
```diff
@@ -561,18 +561,24 @@ class MetadataValidator(Validator):
+        if skip == "":
+            self.skip = None
+        else:
+            self.skip = skip.split(",")
```
```suggestion
        self.check = check.split(",") if check else None
        self.skip = skip.split(",") if skip else None
```

nsoranzo commented on lib/galaxy/tools/parameters/validation.py:
```diff
@@ -561,18 +561,24 @@ class MetadataValidator(Validator):
                    negate=elem.get('negate', 'false'))
 
     def __init__(self, message=None, check="", skip="", negate='false'):
+        if message is None:
```
```suggestion
        if not message:
```

bernt-matthias committed:
Improve MetadataValidation

the `missing_meta` function skipped optional metadata if `skip`
was provided. Note skip only makes sense for optional metadata
since non-optional metadata is validated by the default metadata
validator (which uses check=None and skip=None).

In `missing_meta`: allow for 0, "", and None as valid metadata.

Fix tests for metadata validator:

- use test-data from GALAXY_ROOT/test-data/
- use real (optional) metadata for check and skip

Co-authored-by: Nicola Soranzo <nicola.soranzo@earlham.ac.uk>

bernt-matthias committed:
Simplify handling of None defaults

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

bernt-matthias committed:
extend docs for metadata validator

nsoranzo commented on lib/galaxy/tools/parameters/validation.py:
```diff
@@ -18,7 +18,7 @@
     """Returns test data filename"""
     path, name = os.path.split(__file__)
-    full_path = os.path.join(path, 'test', fname)
+    full_path = os.path.join(path, '..', '..', '..', '..', 'test-data', fname)
```
This breaks the package structure, and therefore the `test_galaxy_packages` tests.

bernt-matthias replied:
True. What would be the simplest solution for this? Create a test data directory in `tools` and link the test data?

nsoranzo replied:
There's already one, with symlinks: lib/galaxy/tools/parameters/test/

bernt-matthias committed:
fix test data location

bernt-matthias committed:
check if metadata is set by comparing to no_value

using the element_is_set function

bernt-matthias committed:
remove no_value from metadata definitions in datatypes

since its the default

nsoranzo commented on lib/galaxy/datatypes/data.py:
```diff
@@ -229,18 +229,18 @@ def missing_meta(self, dataset, check=None, skip=None):
+        for key in check:
+            if key in skip:
+                continue
+            if len(check) == 0 and len(skip) == 0 and dataset.metadata.spec[key].get("optional"):
```
Now that you are reusing `check`, `len(check)` will never be 0 inside the loop.

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -3730,7 +3730,10 @@ In case of ``data_collection`` parameters and
+- ``metadata``: Check for missing metadata. The set of (optional) metadata
+to be checked can be set using either the ``check`` or ``skip`` attribute.
+Note that each data parameter has automatically a metadata validator that checks
+if all non-optional metadata is set, i.e. ``<validator type="metadata/>``.
```
```suggestion
if all non-optional metadata are set, i.e. ``<validator type="metadata/>``.
```

bernt-matthias committed:
reformulate doc

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

bernt-matthias committed:
re-fix missing_meta

bernt-matthias committed:
Merge branch 'fix/metadata-validation' of https://github.com/bernt-matthias/galaxy into fix/metadata-validation

bernt-matthias committed:
improvements for MetadataValidator

- make missing_meta return missing metadata keys / False
- include missing metadata name in MetadataValidator error
- proper messages for failing negated MetadataValidator
- reuse item getter in element_is_set

bernt-matthias committed:
make dbkey metadata optional

because it clearly is optional (in particular because default == no_value)

bernt-matthias committed:
biom2 allways set format and format-version

biom2 sets either format of format_version. since the metadata elements
are not optional we should just set both to the same value

mvdbeek reviewed (approved):
Is this ready ?
